adl/adl.c

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright 2011-2012 Con Kolivas
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 3 of the License, or (at your option)
+ * any later version.  See COPYING for more details.
+ */
+
+#include <stddef.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include "adl_sdk.h"
+#include "adl_functions.h"
+
+#define MAX_GPUDEVICES 16
+
+// declarations in adl_functions.h for these are formatted for dynamic loading
+int ADL_Adapter_AdapterInfo_Get(LPAdapterInfo lpInfo, int iInputSize);
+int ADL_Adapter_ID_Get(int iAdapterIndex, int *lpAdapterID);
+int ADL_Adapter_NumberOfAdapters_Get(int *lpNumAdapters);
+int ADL_Main_Control_Create(ADL_MAIN_MALLOC_CALLBACK callback, int iEnumConnectedAdapters);
+int ADL_Overdrive5_FanSpeed_Get(int iAdapterIndex, int iThermalControllerIndex, ADLFanSpeedValue *lpFanSpeedValue);
+int ADL_Overdrive5_Temperature_Get (int iAdapterIndex, int iThermalControllerIndex, ADLTemperature *lpTemperature);
+
+int getADLInfo(int deviceid, char field[64]);
+
+struct gpu_adapters {
+  int iAdapterIndex;
+  int iBusNumber;
+  int virtual_gpu;
+  int id;
+};
+
+// Memory allocation function
+static void * __stdcall ADL_Main_Memory_Alloc(int iSize)
+{
+  void *lpBuffer = malloc(iSize);
+
+  return lpBuffer;
+}
+
+// Optional Memory de-allocation function
+static void __stdcall ADL_Main_Memory_Free (void **lpBuffer)
+{
+  if (*lpBuffer != NULL) {
+    free (*lpBuffer);
+    *lpBuffer = NULL;
+  }
+}
+
+int getADLFanPercent(int deviceid) {
+  int fanPercent = 0;
+  fanPercent = getADLInfo(deviceid, "fanPercent");
+  return fanPercent;
+}
+
+int getADLTemp(int deviceid) {
+  int temp = 0;
+  temp = getADLInfo(deviceid, "temp");
+  return temp;
+}
+
+int getADLInfo(int deviceid, char field[64]) {
+  int result, i, j, devices = 0, last_adapter = -1, gpu = 0, dummy = 0;
+  int iNumberAdapters;
+  struct gpu_adapters adapters[MAX_GPUDEVICES], vadapters[MAX_GPUDEVICES];
+  bool devs_match = true;
+  ADLBiosInfo BiosInfo;
+  LPAdapterInfo lpInfo = NULL;
+
+  if (ADL_OK != ADL_Main_Control_Create(ADL_Main_Memory_Alloc, 1)) {
+    //printf("ADL_Main_Control_Create failed\n");
```

Comment:
I was planning on cleaning that up as part of #104 but I guess I could do it now.